core/src/main/java/io/undertow/protocols/http2/Http2PushBackParser.java

```diff
@@ -63,28 +64,48 @@ public void parse(ByteBuffer data, Http2FrameHeaderParser headerParser) throws I
             used = rem - dataToParse.remaining();
             if(!isFinished() && remainingData > 0 && used == 0 && dataToParse.remaining() >= remainingData) {
                 if(cnt++ == 100) {
-                    throw UndertowMessages.MESSAGES.parserDidNotMakeProgress();
+                    original = UndertowMessages.MESSAGES.parserDidNotMakeProgress();
                 }
             }
-
+        } catch (Throwable t) {
+            original = t;
         } finally {
             //it is possible that we finished the parsing without using up all the data
             //and the rest is to be consumed by the stream itself
-            if (finished) {
-                data.limit(oldLimit);
-                return;
-            }
-            int leftOver = dataToParse.remaining();
-            if (leftOver > 0) {
-                pushedBackData = new byte[leftOver];
-                dataToParse.get(pushedBackData);
-            } else {
-                pushedBackData = null;
+            try {
+                if (finished) {
+                    data.limit(oldLimit);
+                } else {
+                    int leftOver = dataToParse.remaining();
+                    if (leftOver > 0) {
+                        pushedBackData = new byte[leftOver];
+                        dataToParse.get(pushedBackData);
+                    } else {
+                        pushedBackData = null;
+                    }
+                    data.limit(oldLimit);
+                    remainingData -= used;
+                    if (remainingData == 0) {
+                        finished = true;
+                    }
+                }
+            } catch (Throwable t) {
+                if(original != null) {
+                    original.addSuppressed(t);
+                } else {
+                    original = t;
+                }
             }
-            data.limit(oldLimit);
-            remainingData -= used;
-            if (remainingData == 0) {
-                finished = true;
+            if (original != null) {
+                if (original instanceof RuntimeException) {
```

Comment:
All three 'instanceof' checks are useless .

Reply:
@ropalka Could you explain the issue, please? The compiler refuses to throw the (undeclared) Throwable

Reply:
You're right of course!